crates/turborepo-lib/src/boundaries.rs

```diff
@@ -0,0 +1,443 @@
+    package_json::PackageJson,
+};
+
+use crate::run::Run;
+
+#[derive(Clone, Debug, Error, Diagnostic)]
+pub enum BoundariesDiagnostic {
+    #[error(
+        "importing from a type declaration package, but import is not declared as a type-only \
+         import"
+    )]
+    #[help("add `type` to the import declaration")]
+    NotTypeOnlyImport {
+        import: String,
+        #[label("package imported here")]
+        span: SourceSpan,
+        #[source_code]
+        text: Arc<NamedSource>,
+    },
+    #[error("cannot import package `{name}` because it is not a dependency")]
+    PackageNotFound {
+        name: String,
+        #[label("package imported here")]
+        span: SourceSpan,
+        #[source_code]
+        text: Arc<NamedSource>,
+    },
+    #[error("cannot import file `{import}` because it leaves the package")]
+    ImportLeavesPackage {
+        import: String,
+        #[label("file imported here")]
+        span: SourceSpan,
+        #[source_code]
+        text: Arc<NamedSource>,
+    },
+    #[error("failed to parse file {0}")]
+    ParseError(AbsoluteSystemPathBuf, swc_ecma_parser::error::Error),
+}
+
+#[derive(Debug, Error, Diagnostic)]
+pub enum Error {
+    #[error("file `{0}` does not have a parent directory")]
+    NoParentDir(AbsoluteSystemPathBuf),
+    #[error(transparent)]
+    Path(#[from] turbopath::PathError),
+    #[error(transparent)]
+    Lockfiles(#[from] turborepo_lockfiles::Error),
+    #[error(transparent)]
+    GlobWalk(#[from] globwalk::WalkError),
+    #[error("failed to read file: {0}")]
+    FileNotFound(AbsoluteSystemPathBuf),
+}
+
+static PACKAGE_NAME_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(r"^(@[a-z0-9-~][a-z0-9-._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$").unwrap()
+});
+
+pub struct BoundariesResult {
+    pub source_map: Arc<SourceMap>,
+    pub diagnostics: Vec<BoundariesDiagnostic>,
+}
+
+impl BoundariesResult {
+    pub fn is_ok(&self) -> bool {
+        self.diagnostics.is_empty()
+    }
+
+    pub fn emit(&self) {
+        let handler = Handler::with_tty_emitter(
+            ColorConfig::Auto,
+            true,
+            false,
+            Some(self.source_map.clone()),
+        );
+
+        for diagnostic in self.diagnostics.clone() {
+            match diagnostic {
+                BoundariesDiagnostic::ParseError(_, e) => {
+                    e.clone().into_diagnostic(&handler).emit();
+                }
+                e => {
+                    eprintln!("{:?}", Report::new(e.to_owned()));
+                }
+            }
+        }
+    }
+}
+
+impl Run {
+    pub async fn check_boundaries(&self) -> Result<BoundariesResult, Error> {
+        let packages = self.pkg_dep_graph().packages();
+        let repo = Repository::discover(&self.repo_root()).ok().map(Mutex::new);
+        let mut diagnostics = vec![];
+        let source_map = SourceMap::default();
+        for (package_name, package_info) in packages {
```

Comment:
We could consider parallelizing this, but it's not *that* slow right now (7 seconds for running in `turborepo` on release mode)